drivers/counter/counter_ll_stm32_timer.c

```diff
@@ -0,0 +1,623 @@
+{
+	const struct counter_stm32_config *cfg = dev->config;
+	uint32_t pending = 0;
+
+	switch (counter_get_num_of_channels(dev)) {
+	case 4:
+		pending |= LL_TIM_IsActiveFlag_CC4(cfg->timer);
+	case 3:
+		pending |= LL_TIM_IsActiveFlag_CC3(cfg->timer);
+	case 2:
+		pending |= LL_TIM_IsActiveFlag_CC2(cfg->timer);
+	case 1:
+		pending |= LL_TIM_IsActiveFlag_CC1(cfg->timer);
+	}
+
+	return !!pending;
+}
+
+/**
+ * Obtain timer clock speed.
+ *
+ * @param pclken  Timer clock control subsystem.
+ * @param tim_clk Where computed timer clock will be stored.
+ *
+ * @return 0 on success, error code otherwise.
+ *
+ * This function is ripped from the PWM driver; TODO handle code duplication.
+ */
+static int counter_stm32_get_tim_clk(const struct stm32_pclken *pclken, uint32_t *tim_clk)
+{
+	int r;
+	const struct device *clk;
+	uint32_t bus_clk, apb_psc;
+
+	clk = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+
+	r = clock_control_get_rate(clk, (clock_control_subsys_t *)pclken,
+				   &bus_clk);
+	if (r < 0) {
+		return r;
+	}
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	if (pclken->bus == STM32_CLOCK_BUS_APB1) {
+		apb_psc = STM32_D2PPRE1;
+	} else {
+		apb_psc = STM32_D2PPRE2;
+	}
+#else
+	if (pclken->bus == STM32_CLOCK_BUS_APB1) {
+		apb_psc = STM32_APB1_PRESCALER;
+	}
+#if !defined(CONFIG_SOC_SERIES_STM32F0X) && !defined(CONFIG_SOC_SERIES_STM32G0X)
+	else {
+		apb_psc = STM32_APB2_PRESCALER;
+	}
+#endif
+#endif
+
+#if defined(RCC_DCKCFGR_TIMPRE) || defined(RCC_DCKCFGR1_TIMPRE) || \
+	defined(RCC_CFGR_TIMPRE)
+	/*
+	 * There are certain series (some F4, F7 and H7) that have the TIMPRE
+	 * bit to control the clock frequency of all the timers connected to
+	 * APB1 and APB2 domains.
+	 *
+	 * Up to a certain threshold value of APB{1,2} prescaler, timer clock
+	 * equals to HCLK. This threshold value depends on TIMPRE setting
+	 * (2 if TIMPRE=0, 4 if TIMPRE=1). Above threshold, timer clock is set
+	 * to a multiple of the APB domain clock PCLK{1,2} (2 if TIMPRE=0, 4 if
+	 * TIMPRE=1).
+	 */
+
+	if (LL_RCC_GetTIMPrescaler() == LL_RCC_TIM_PRESCALER_TWICE) {
+		/* TIMPRE = 0 */
+		if (apb_psc <= 2u) {
+			LL_RCC_ClocksTypeDef clocks;
+
+			LL_RCC_GetSystemClocksFreq(&clocks);
+			*tim_clk = clocks.HCLK_Frequency;
+		} else {
+			*tim_clk = bus_clk * 2u;
+		}
+	} else {
+		/* TIMPRE = 1 */
+		if (apb_psc <= 4u) {
+			LL_RCC_ClocksTypeDef clocks;
+
+			LL_RCC_GetSystemClocksFreq(&clocks);
+			*tim_clk = clocks.HCLK_Frequency;
+		} else {
+			*tim_clk = bus_clk * 4u;
+		}
+	}
+#else
+	/*
+	 * If the APB prescaler equals 1, the timer clock frequencies
+	 * are set to the same frequency as that of the APB domain.
+	 * Otherwise, they are set to twice (×2) the frequency of the
+	 * APB domain.
+	 */
+	if (apb_psc == 1u) {
+		*tim_clk = bus_clk;
+	} else {
+		*tim_clk = bus_clk * 2u;
+	}
+#endif
+
+	return 0;
+}
+
+static int counter_stm32_init_timer(const struct device *dev)
+{
+	struct counter_stm32_config *cfg =
+		(struct counter_stm32_config *)DEV_CFG(dev);
```

Comment:
please, don't use this pattern, just do `struct counter_stm32_config *cfg = dev->config;`

Reply:
This is necessary to cast away constness for setting `freq` on runtime initialization.

Reply:
Whoa, casting away const for device configuration data is not a good idea. The device initialization macro will put the entire configuration structure into flash on STM32, so writes should fail. Did you mean to put this into the data structure instead of the config structure?

Reply:
For STM32, it seems that the frequency needs to be determined at runtime, which means it can't be initialized the normal way (in the struct definition). So in an earlier thread with @ABOSTM, we decided to set it this way. There additionally seems to be precedent for this practice in other counter drivers, namely `drivers/counter/counter_mcux_ctimer.c` and `drivers/counter/counter_imx_epit.c`.

Reply:
@kentjhall If you need to set `freq` in driver operation, please put freq in `counter_stm32_data`, that's why these structs are there.

Reply:
The intent is for `counter_get_frequency` to work properly. My understanding is that `freq` in `struct counter_config_info` must be set correctly for this:

```
static inline uint32_t z_impl_counter_get_frequency(const struct device *dev)
{
	const struct counter_config_info *config =
			(const struct counter_config_info *)dev->config;

	return config->freq;
}
```

Am I missing something?

Reply:
@nordic-krch, would you mind suggesting how to proceed on this point ?

Reply:
@nordic-krch @erwango Just bumping this since it's been a couple weeks, any update?

Reply:
In that case maybe it's worth adding optional `get_freq` function pointer to counter api structure and have a call like:

```
static inline uint32_t z_impl_counter_get_frequency(const struct device *dev)
{
       if (api->get_freq) {
          return api->get_freq(dev);
       }

	const struct counter_config_info *config =
			(const struct counter_config_info *)dev->config;

	return config->freq;
}
```
That would solve it for case when frequency cannot be determine at runtime.
What do you think?

Reply:
That sounds reasonable to me. Would definitely remedy the situation here.

Reply:
Is someone to work on a separate PR for that @nordic-krch? I wouldn't mind doing myself as well, seems like it would be very quick.

Reply:
Ok @nordic-krch @erwango, this has been updated to make use of my merged `get_freq` PR.